### PR TITLE
add trx for rel-deletion

### PR DESCRIPTION
**Pull request by acquamarin**
1. This PR adds the transaction logic for rel-deletion.
2. Refactors the prepare-commiting logic for lists:
Instead of commiting each list at a time, we commit each relTable at a time.

### CODE REVIEW

File: src/storage/store/rel_table.cpp [hunk 1]
```diff
@@ -49,10 +50,107 @@ vector<AdjColumn*> RelTable::getAdjColumnsForNodeTable(table_id_t tableID) {
...
+                        // Note: An empty listUpdates can exist for a nodeOffset, because we don't
+                        // fix the listUpdates, listUpdatesPerNode and ListUpdatesPerChunk indices
+                        // after we insert or delete a rel. For example: a user inserts 1 rel to
+                        // nodeOffset1, and then deletes that rel. We will end up getting an empty
+                        // listUpdates for nodeOffset1.
+                        if (!listUpdates.hasUpdates()) {
+                            continue;
+                        }
+                        if (listUpdates.emptyListInPersistentStore) {
+                            listsUpdateIterators[0]->updateList(nodeOffset,
```
Comment (semihsalihoglu-uw) [inline src/storage/store/rel_table.cpp:82]: As a second comment I'm now debating if this first `if clause` is justified. Why don't you just have 2 cases: 

1) if the list is a large list and there are only updates;
2) else

So you don't differentiate between whether the persistent list was empty or not and just call iterator->updateList(...), which will check whether to fill from the persistent store or not. I'm assuming updateList() should actually just work even if the persistent store is empty. It would simplify the code.

Given that you do this, I have a follow up question: Aside from here do we ever use emptyListInPersistentStore? If we remove this if case, it may be unused and if so, we should remove it too.

File: test/runner/e2e_delete_rel_test.cpp [hunk 2]
```diff
@@ -19,104 +19,316 @@ class DeleteRelTest : public DBTest {
...
+        if (isCommit) {
+            expectedResult = vector<string>{"0", "1", "2", "3", "4", "5", "6", "7", "8", "9", "50"};
+        } else {
+            for (auto i = 0u; i <= 50; i++) {
+                expectedResult.push_back(to_string(i));
+            }
+        }
+        ASSERT_EQ(TestHelper::convertResultToString(
+                      *conn->query("MATCH (a:animal)-[e:knows]->(p:person) RETURN e.length"),
+                      true /* checkOutputOrder */),
```
Comment (semihsalihoglu-uw) [inline test/runner/e2e_delete_rel_test.cpp:160]: Why are you checking the output order in a query that does not have ORDER BY?

File: test/runner/e2e_delete_rel_test.cpp [hunk 3]
```diff
@@ -19,104 +19,316 @@ class DeleteRelTest : public DBTest {
...
+    }
 
-    auto actualResult = TestHelper::convertResultToString(
-        *conn->query("MATCH (:animal)-[e:knows]->(:person) return e.length"),
-        true /* checkOutputOrder */);
-    // After insertion and deletion, we are expected to see only one new rel: animal52->person0.
-    vector<string> expectedResult;
-    for (auto i = 0; i < NUM_RELS_FOR_ANIMAL_KNOWS_PERSON; i++) {
-        expectedResult.push_back(to_string(i));
+    void deleteMultipleRels(bool isCommit, TransactionTestType transactionTestType) {
```
Comment (semihsalihoglu-uw) [inline test/runner/e2e_delete_rel_test.cpp:143]: The previous tests above, e.g., deleteRelsFromSmall/LargList or deleteRelsFromUpdateStore  also delete multiple rels? What exactly is this testing? If this test doesn't add much, let's remove it.

File: test/runner/e2e_delete_rel_test.cpp [hunk 4]
```diff
@@ -19,104 +19,316 @@ class DeleteRelTest : public DBTest {
...
+                    "MATCH (a:animal)-[e:knows]->(p:person) WHERE a.ID >= 40 RETURN e.length"),
+                true /* checkOutputOrder */),
+            expectedResult);
     }
-    expectedResult.push_back(to_string(52));
-    ASSERT_EQ(actualResult, expectedResult);
+
+    void deleteLargeNumRelsFromLargeList(bool isCommit, TransactionTestType transactionTestType) {
+        conn->beginWriteTransaction();
+        conn->query(
```
Comment (semihsalihoglu-uw) [inline test/runner/e2e_delete_rel_test.cpp:189]: Is this deleting all rels or almost all rels? Maybe remove this test and add two tests that test the edge cases of deleting all rels from a small and large list.

**Review by semihsalihoglu-uw — APPROVED** [review]

File: src/include/storage/storage_structure/lists/lists.h [hunk 5]
```diff
@@ -116,23 +116,26 @@ class Lists : public BaseColumnOrList {
...
-        uint64_t numElementsInPersistentStore, InMemList& inMemList);
+        uint64_t numElfementsInPersistentStore, InMemList& inMemList,
+        const unordered_set<uint64_t>& deletedRelOffsetsInList);
     uint64_t getNumElementsInPersistentStore(
         TransactionType transactionType, node_offset_t nodeOffset);
     void initListReadingState(
         node_offset_t nodeOffset, ListHandle& listHandle, TransactionType transactionType);
     unique_ptr<InMemList> getInMemListWithDataFromUpdateStoreOnly(
         node_offset_t nodeOffset, vector<uint64_t>& insertedRelsTupleIdxInFT);
+    // This function writes the persistent store data (except those deleted rels) and update store
```
[the remark this answers is not in the document]
Comment (acquamarin) [reply src/include/storage/storage_structure/lists/lists.h:134]: Changed `deleted rels` -> `skipping over the deleted rels`

File: src/include/storage/storage_structure/lists/lists_update_store.h [hunk 6]
```diff
@@ -17,7 +17,15 @@ using namespace processor;
 using namespace catalog;
 
 struct ListUpdates {
+public:
     ListUpdates() : emptyListInPersistentStore{false} {}
+
+    inline bool hasUpdates() const {
+        return emptyListInPersistentStore || !insertedRelsTupleIdxInFT.empty() ||
```
[the remark this answers is not in the document]
Comment (acquamarin) [reply src/include/storage/storage_structure/lists/lists_update_store.h:24]: The purpose of the `hasUpdates()` is to check whether we need to update the list/header/metadata for the current nodeOffset. 
We want `hasUpdates` to return true if we want to update the list/header/metadata for a nodeOffset. When we insert a new node, we set the `emptyListIPersistentStore` to true in the listUpdates for that new node. During commiting, we need to fix the header for that newly inserted node. As a result, we need `hasUpdates`  to return true in order to fix the header.

File: src/storage/storage_structure/lists/lists.cpp [hunk 7]
```diff
@@ -137,49 +125,54 @@ unique_ptr<InMemList> Lists::getInMemListWithDataFromUpdateStoreOnly(
...
-                auto numElementsInPersistentStore = getNumElementsFromListHeader(nodeOffset);
-                fillInMemListsFromPersistentStore(
-                    cursorAndMapper, numElementsInPersistentStore, *inMemList);
-                listsUpdateStore->readInsertionsToList(
-                    storageStructureIDAndFName.storageStructureID.listFileID,
-                    updatedNodeOffsetItr->second.insertedRelsTupleIdxInFT, *inMemList,
-                    numElementsInPersistentStore, getDiskOverflowFileIfExists(), dataType,
-                    getNodeIDCompressionIfExists());
-                listsUpdateIterator.updateList(nodeOffset, *inMemList);
+unique_ptr<InMemList> Lists::writeToInMemList(node_offset_t nodeOffset,
```
[the remark this answers is not in the document]
Comment (acquamarin) [reply src/storage/storage_structure/lists/lists.cpp:128]: Sure, moved `fillInMemListsFromPersistentStore` below `writeToInMemList`

File: src/storage/storage_structure/lists/lists.cpp [hunk 8]
```diff
@@ -137,49 +125,54 @@ unique_ptr<InMemList> Lists::getInMemListWithDataFromUpdateStoreOnly(
...
+        numElementsInPersistentStore - deletedRelOffsetsForList.size(),
+        getDiskOverflowFileIfExists(), dataType, getNodeIDCompressionIfExists());
+    return inMemList;
+}
+
+void Lists::fillInMemListsFromFrame(InMemList& inMemList, uint8_t* frame, uint64_t elemPosInPage,
+    uint64_t numElementsToReadInCurPage, const unordered_set<uint64_t>& deletedRelOffsetsInList,
+    uint64_t numElementsRead, uint64_t& nextPosToWriteToInMemList) {
+    auto nullBufferInPage = (uint64_t*)(frame + numElementsPerPage * elementSize);
+    auto frameData = frame + elemPosInPage * elementSize;
```
[the remark this answers is not in the document]
Comment (acquamarin) [reply src/storage/storage_structure/lists/lists.cpp:151]: Created a new function     
```
inline const uint8_t* getElemFrameInPage(
        const uint8_t* pageFrame, uint64_t elemPosInPage) const {
        return pageFrame + elemPosInPage * elementSize;
    }
```

File: test/runner/e2e_delete_rel_test.cpp (hunk 2 above)
Comment (acquamarin) [reply test/runner/e2e_delete_rel_test.cpp:160]: The reason is that if the `checkOutputOrder` is false, the testHelper will sort the actual test results. 
Anyways, i changed checkOutputOrder to false, and sort the expectedResult when comparing with the actualResult.

File: test/runner/e2e_delete_rel_test.cpp (hunk 3 above)
Comment (acquamarin) [reply test/runner/e2e_delete_rel_test.cpp:143]: 1. deleteRelsFromSmallList is to test the behaviour of deleting multiple rels from a small list. Specifically, i want to test whether the CSR structure has been fixed properly after each deletion.
2. deleteRelsFromLargelist is to test the behaviour of deleting multiple rels from a large list. Specifically, i want to test whether the largelist has been fixed properly after each deletion.
3. deleteRelsFromUpdateStore is quite different from deleting from persistent store. When we want to delete rels from updateStore, we can simply remove the rels from the insertedRelTupleIdxes of updateStore.

File: test/runner/e2e_delete_rel_test.cpp (hunk 4 above)
Comment (acquamarin) [reply test/runner/e2e_delete_rel_test.cpp:189]: This test is to test deleting large number of rels from a large list.
Addded two tests for deleting all rels from small/large list.

File: src/storage/storage_structure/lists/lists.cpp [hunk 9]
```diff
@@ -137,49 +125,54 @@ unique_ptr<InMemList> Lists::getInMemListWithDataFromUpdateStoreOnly(
...
+            NullMask::copyNullMask(nullBufferInPage, elemPosInPage, inMemList.getNullMask(),
+                nextPosToWriteToInMemList, numElementsToReadInCurPage);
+        }
+        nextPosToWriteToInMemList += numElementsToReadInCurPage;
+    } else {
+        // If we have some deleted rels, we should check whether each rel has been deleted
+        // or not before copying to inMemList.
+        for (auto i = 0u; i < numElementsToReadInCurPage; i++) {
+            auto relOffsetInList = numElementsRead + i;
+            if (!deletedRelOffsetsInList.contains(relOffsetInList)) {
```
[the remark this answers is not in the document]
Comment (acquamarin) [reply src/storage/storage_structure/lists/lists.cpp:166]: Opened an issue for this improvement: https://github.com/kuzudb/kuzu/issues/1130

File: src/storage/storage_structure/lists/lists.cpp [hunk 10]
```diff
@@ -374,5 +367,35 @@ void RelIDList::setDeletedRelsIfNecessary(Transaction* transaction, ListSyncStat
     }
 }
 
+unordered_set<uint64_t> RelIDList::getDeletedRelOffsetsInListForNodeOffset(
```
[the remark this answers is not in the document]
Comment (acquamarin) [reply src/storage/storage_structure/lists/lists.cpp:370]: opened an issue for this improvement: #1130

File: src/storage/store/rel_table.cpp (hunk 1 above)
Comment (acquamarin) [reply src/storage/store/rel_table.cpp:82]: As we discussed, let's keep the `listsUpdates.emptyListInPersistentStore` branch here since the logic is quite different from the newly inserted nodes and other nodes:
```
If the node is newly inserted, we should only read from the updateStore.
If the node is not newly inserted, we should both read from the persistent and update store.
```
btw: renamed the `emptyListInPersistentStore` -> `newlyAddedNode`

File: src/storage/store/rel_table.cpp [hunk 11]
```diff
@@ -49,10 +50,107 @@ vector<AdjColumn*> RelTable::getAdjColumnsForNodeTable(table_id_t tableID) {
...
+                        auto& listUpdates = updatedNodeOffsetItr->second;
+                        // Note: An empty listUpdates can exist for a nodeOffset, because we don't
+                        // fix the listUpdates, listUpdatesPerNode and ListUpdatesPerChunk indices
+                        // after we insert or delete a rel. For example: a user inserts 1 rel to
+                        // nodeOffset1, and then deletes that rel. We will end up getting an empty
+                        // listUpdates for nodeOffset1.
+                        if (!listUpdates.hasUpdates()) {
+                            continue;
+                        }
+                        if (listUpdates.emptyListInPersistentStore) {
```
[the remark this answers is not in the document]
Comment (acquamarin) [reply src/storage/store/rel_table.cpp:81]: I think we already have a comment for the large list optimization.
The `if (listUpdates.emptyListInPersistentStore)` is a very special branch which is designed for newly inserted nodes.
